File: leet_code/go/maximum_binary_tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */


func constructMaximumBinaryTree(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}
	mid := findIndexOfMax(nums)
	left := nums[0:mid]
	right := nums[(mid + 1):len(nums)]
	root := TreeNode{ Val: nums[mid] }
	root.Left = constructMaximumBinaryTree(left)
	root.Right = constructMaximumBinaryTree(right)
	return &root
}

func findIndexOfMax(nums []int) int {
	max := 0
	for i, val := range nums {
		if val > nums[max] {
			max = i
		}
	}
	return max
}

File: leet_code/ruby/maximum_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {Integer[]} nums
# @return {TreeNode}
def construct_maximum_binary_tree(nums)
	return if nums.empty?
	mid = nums.each_with_index.max[1]
	root = TreeNode.new(nums[mid])
	root.left = construct_maximum_binary_tree(nums[0...mid])
	root.right = construct_maximum_binary_tree(nums[(mid + 1)...nums.size])
	root
end